pkg/util/loop/loop.go

```diff
@@ -168,82 +169,48 @@ func (loop *Device) shareLoop(imageIno, imageDev uint64, mode int, number *int)
 // For most failures with loopN, it will try loopN+1, continuing up to loop.MaxLoopDevices.
 // When setting loop device status, some kernel may return EAGAIN, this function would sync
 // workaround this error.
-func (loop *Device) attachLoop(image *os.File, mode int, number *int) error {
-	// Because we hold a lock on /dev here, avoid delayed retries inside this function,
-	// as it could impact parallel startup of many instances of Singularity or
-	// other programs.
-	fd, err := lock.Exclusive("/dev")
-	if err != nil {
-		return err
-	}
-	defer lock.Release(fd)
-
+func (loop *Device) attachLoop(imageFd uintptr, imageInfo *syscall.Stat_t, mode int, number *int) error {
 	releaseDevice := func(fd int, clear bool) {
 		if clear {
 			syscall.Syscall(syscall.SYS_IOCTL, uintptr(fd), CmdClrFd, 0)
 		}
 		syscall.Close(fd)
 	}
 
+	retryFn := getRetryStatusFn(imageFd, imageInfo)
+
 	for device := 0; device < loop.MaxLoopDevices; device++ {
 		*number = device
-		clearDevice := false
 
 		// Try to open the loop device, creating the device node if needed
-		loopFd, err := openLoopDev(device, mode, true)
+		loopFd, releaseLock, err := openLoopDev(device, mode, true, loop.Shared)
 		if err != nil {
-			sylog.Debugf("couldn't openLoopDev loop device %d: %v", device, err)
+			sylog.Debugf("Couldn't open loop device %d: %v", device, err)
 			return err
 		}
 
```

Comment:
@cclerget what about `defer releaseDevice...`

Reply:
stacking defer inside loop is not ideal here as it would lead to side effects like hitting the file descriptor limit